test/e2e/autoscale_hpa_test.go

```diff
@@ -0,0 +1,205 @@
+	primeNum              = 500000
+)
+
+func TestHPAAutoscaleUpDownUp(t *testing.T) {
+	ctx := setupHPASvc(t, autoscaling.HPA, autoscaling.CPU, cpuTarget)
+	test.EnsureTearDown(t, ctx.Clients(), ctx.Names())
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+	assertScaleDownToOne(ctx)
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+}
+
+func setupHPASvc(t *testing.T, class, metric string, target int) *TestContext {
+	t.Helper()
+	clients := Setup(t)
+
+	t.Log("Creating a new Route and Configuration")
+	names := &test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   autoscaleTestImageName,
+	}
+	resources, err := v1test.CreateServiceReady(t, clients, names,
+		[]rtesting.ServiceOption{
+			rtesting.WithConfigAnnotations(map[string]string{
+				autoscaling.ClassAnnotationKey:    class,
+				autoscaling.MetricAnnotationKey:   metric,
+				autoscaling.TargetAnnotationKey:   strconv.Itoa(target),
+				autoscaling.MinScaleAnnotationKey: fmt.Sprintf("%d", minimumReplicas),
+				autoscaling.MaxScaleAnnotationKey: fmt.Sprintf("%d", int(maxPods)),
+			}), rtesting.WithResourceRequirements(corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("30m"),
+					corev1.ResourceMemory: resource.MustParse("20Mi"),
+				},
+			}),
+		}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	if _, err := pkgTest.WaitForEndpointState(
+		context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		names.URL,
+		v1test.RetryingRouteInconsistency(spoof.MatchesAllOf(spoof.IsStatusOK)),
+		"CheckingEndpointAfterCreate",
+		false,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
+	); err != nil {
+		t.Fatalf("Error probing %s: %v", names.URL.Hostname(), err)
+	}
+
+	return &TestContext{
+		t:           t,
+		logf:        t.Logf,
+		clients:     clients,
+		names:       names,
+		resources:   resources,
+		targetValue: target,
+		metric:      metric,
+	}
+}
+func assertHPAAutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) {
+	ctx.t.Helper()
+	wait := autoscaleHPAUpToNumPods(ctx, curPods, targetPods, done, quick)
+	if err := wait(); err != nil {
+		ctx.t.Fatal(err)
+	}
+}
+
+func autoscaleHPAUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) func() error {
+	ctx.t.Helper()
+
+	stopChan := make(chan struct{})
+	var grp errgroup.Group
+	grp.Go(func() error {
+		return generateTrafficAtFixedRPSWithCPULoad(ctx, int(targetPods*5), stopChan)
+	})
+
+	grp.Go(func() error {
+		defer close(stopChan)
+		return checkPodScale(ctx, targetPods, minPods, maxPods, done, quick)
+	})
+
+	return grp.Wait
+}
+
+func generateTrafficAtFixedRPSWithCPULoad(ctx *TestContext, rps int, stopChan chan struct{}) error {
```

Comment:
I wonder if it'd make sense for us to control the concurrency rather than RPS here, to have tighter control over how much load we're generating exactly. Using RPS, we have no control over the peak CPU usage we're inducing, right?

Reply:
Yes I thought about this and rps felt more natural to quickly hit a high peak so that we get the cpu load we need. 
There is no control but that is why the numbers are low (25rps) so it does not crash the pods (it happened with high numbers of either prime calculation or number of rps).